File: best_time_to_buy_and_sell_stock.go
// @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
// @ Copyright (c) Michael Leahcim                                                      @
// @ You can find additional information regarding licensing of this work in LICENSE.md @
// @ You must not remove this notice, or any other, from this software.                 @
// @ All rights reserved.                                                               @
// @@@@@@ At 2019-03-31 08:52 <thereisnodotcollective@gmail.com> @@@@@@@@@@@@@@@@@@@@@@@@
package main

// Solving: https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iii/

// Problem definition:
// for a1...aN choose four numbers a,b,c,d, where  index(a)<index(b)<index(c)<index(d) in such a way
//   that (b-a) + (c-d) is maximum

// Can we reuse solution for X-1?
//  [1,3,1,3] => 4 (3-1) + (3-1)
//  solution([1]) => 0   (hardcoded)
//  solution([1,3]) => 2 (hardcoded)
//  solution([1,3,1]) => 2 (hardcoded)
//  solution([1,3,1,3]) => 4
//  solution([1,3,1,3,5]) => 6
//  solution([(1),(3),(1),3,(5),1]) => 6
//  solution([(1),3,1,3,(5),(1),(7)]) => 10

// solution([1,3,2,4]) => 4
// solution matrix
// [0, 2,  1,  3
//  N, 0, -1,  1
//  N, N,  0,  2
// 	N, N,  N,  0 ]

// 2,3
//   1
// 4
// [0, 2,  2,  3
//  N, 0,  2,  2
//  N, N,  2,  2
// 	N, N,  N,  2 ]

// solution([3,3,5,0,0,3,1,4])
// [0  0  2 -3 -3  0 -2  1]
// [N  0  2 -3 -3  0 -2  1]
// [N  N  0 -5 -5 -2 -4 -1]
// [N  N  N  0  0  3  1  4]
// [N  N  N  N  0  3  1  4]
// [N  N  N  N  N  0 -2  1]
// [N  N  N  N  N  N  0  3]
// [N  N  N  N  N  N  N  0]

// So, solution_matrix[bought_day][sold_day] = difference
// As per the exampple:
//  Buy on day 4 (price = 0) and sell on day 6 (price = 3),  solution_matrix[4-1][6-1] = 3
//  Then buy on day 7 (price = 1) and sell on day 8 (price = 4), solution_matrix[7-1][8-1] = 3

// What is the solution?
//  solution_matrix[x][y] + solution_matrix[x+x1][y+y1] = maximum
//  solution_matrix[x-x1][y-y1] + solution_matrix[x][y] = maximum where x1,y1 > 0

// So, maxmatrix[x][y] = max(maxmatrix[x-1],maxmatrix[x-1][y-1],maxmatrix[y-1])

// [0  0  2  2  2  2  2  2]
// [N  0  2  2  2  2  2  2]
// [N  N  2  2  2  2  2  2]
// [N  N  N  2  2  3  3  4]
// [N  N  N  N  2  3  3  4]
// [N  N  N  N  N  3  3  4]
// [N  N  N  N  N  N  3  4]
// [N  N  N  N  N  N  N  4]

type best_time_to_buy_and_sell_stock_ struct {
}

var best_time_to_buy_and_sell_stock = best_time_to_buy_and_sell_stock_{}

func (b *best_time_to_buy_and_sell_stock_) GenerateDifferenceMatrix(prices []int) [][]int {
	fs := len(prices)
	result := make([][]int, fs)
	for i := 0; i < fs; i++ {
		pointdiff := make([]int, fs)
		ps := prices[i]
		for j := 0; j < fs; j++ {
			if j < i {
				pointdiff[j] = 0x7FF0000000000000
			} else {
				pointdiff[j] = prices[j] - ps
			}
		}
		result[i] = pointdiff
	}
	return result
}
func (b *best_time_to_buy_and_sell_stock_) MaxProfitMatrix(prices []int) [][]int {
	bo, _ := b.MaxProfit(prices)
	return bo
}
func (b *best_time_to_buy_and_sell_stock_) MaxProfitValue(prices []int) int {
	_, bo := b.MaxProfit(prices)
	return bo
}

// naive implementation
func (b *best_time_to_buy_and_sell_stock_) MaxProfit(prices []int) ([][]int, int) {
	maxmatrix := b.GenerateDifferenceMatrix(prices)
	fs := len(prices)
	best := 0
	for i := 0; i < fs; i++ {
		for j := i; j < fs; j++ {
			right := maxmatrix[i][j]
			lessbest := 0
			for i1 := 0; i1 < i; i1++ {
				for j1 := j; j1 < j; j1++ {
					if lessbest < maxmatrix[i1][j1] {
						lessbest = maxmatrix[i1][j1]
					}
				}
			}
			if right+lessbest > best {
				best = right + lessbest
			}
		}
	}
	return maxmatrix, best

}

func (b *best_time_to_buy_and_sell_stock_) MaxProfitLater(prices []int) ([][]int, int) {
	maxmatrix := b.GenerateDifferenceMatrix(prices)
	best := 0

	fs := len(prices)
	for i := 0; i < fs; i++ {
		for j := i; j < fs; j++ {
			comp := 0
			cur := maxmatrix[i][j]
			// if i > 0 {
			// 	comp = maxmatrix[i-1][j]
			// }

			// if j > i {
			// 	s := maxmatrix[i][j-1]
			// 	if comp < s {
			// 		comp = s
			// 	}
			// }

			if i > 0 && j > i {
				s := maxmatrix[i-1][j-1]
				if comp < s {
					comp = s
				}
				if cur+s > best {
					best = cur + s
				}
			}
			if comp > cur {
				maxmatrix[i][j] = comp
			}
		}
	}
	return maxmatrix, best
}

func maxProfit(prices []int) int {
	return best_time_to_buy_and_sell_stock.MaxProfitValue(prices)
}

File: python/best_time_to_buy_and_sell_stock.py
# @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
# @ Copyright (c) Michael Leahcim                                                      @
# @ You can find additional information regarding licensing of this work in LICENSE.md @
# @ You must not remove this notice, or any other, from this software.                 @
# @ All rights reserved.                                                               @
# @@@@@@ At 2019-04-13 19:25 <thereisnodotcollective@gmail.com> @@@@@@@@@@@@@@@@@@@@@@@@

# Solving: https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iii/
# [3,3,5,0,0,3,1,4]

## Calculate difference matrix

## Difference matrix

## matrix[buyDay][sellDay] = profit

## What is the new problem definition?
### Find such buyDay and sellDay and  buyDay2, sellDay2
###   where their sum is maximum
###   and, buyDay2 >= sellDay

### Approaches?
#### We can find the best solution on the left and on the right for every X_n in input
####   Just iterate over it, and calculate the sum of the given input, choose the best

#### Now, how we can find the best solution for up to N?
###### Can we take the best solution for N-1?
###### Input is: [1,3,2,4]
######   Solution(1)   =   = max(solution(),(1-1))
######   Solution(1,3)     = max(solution(1),3-1,3-3)
######   Solution(1,3,2)   = max(solution(1,3),2-1,2-3,2-2)
######   Solution(1,3,2,4) = max(solution(1,3,2),4-1,4-3,4-2,4-4)

#### Now, we should repeat this process backwards, to find the best solution on the right
#    side of the input
#    Solution(4)   = max(solution(),(4-4))
#    Solution(2,4) = max(solution(4),2-2,4-2)
#    Solution(3,2,4) = max(solution(2,4),3-3,2-3,4-3)
#    Solution(1,3,2,4) = max(solution(3,2,4),1-1,3-1,2-1,4-1)

#### Now, how to optimize?
# I don't like repeated subtractions. On large data size it will take a lot of time
# We can infer subtractions by previous, like this:
##   Solution(1)   =   = max(solution(),(1-1))
##   Solution(1,3)     = max(solution(1),(1-1)+(3-1),3-3)
##   Solution(1,3,2)   = max(solution(1,3),(1-1)+(3-1)+(2-3),(3-3)+(2-3),2-2)
##   Solution(1,3,2,4) = e.t.c.

### No no no, there is an easier way
##  If we find min from input for every step from left and right?
##    [1,3,2,4] => [1,1,1,1]
##  And then, we just subtract that minimum from every point

class Solution(object):
  def calculateMatrix(self,input):
    return [[item2-item if index2>=index else None for index2,item2 in enumerate(input)]
            for index,item in enumerate(input)]
  
  def calculateBestForward(self,input):
    prev     = 0
    yield prev
    
    if len(input) == 0:
      return
    minimum  = input[0]
    
    for index,item in enumerate(input[1:]):
      index = index+1
      prev = max(item-minimum,prev)
      if minimum > item:
        minimum = item
      yield prev
      
  def calculateBestBackward(self,input):
    prev = 0
    yield prev

    if len(input) == 0:
      return
    maximum  = input[len(input)-1]
    
    for index in xrange(len(input)-2,-1,-1):
      item = input[index]
      prev = max(maximum-item,prev)
      if item > maximum:
        maximum = item
      yield prev
  
  def maxProfit(self, prices):
    """
    :type prices: List[int]
    :rtype: int
    """
    best = 0
    for forward,backward in zip(self.calculateBestForward(prices),reversed(list(self.calculateBestBackward(prices)))):
      if forward+backward > best:
        best = forward+backward
    return best      

